### PR TITLE
Reset cursor at dedent after whitespace

### PR DESCRIPTION
**Pull request by dhruvmanila**
## Summary

This PR fixes a bug to reset the cursor before emitting the `Dedent` token which had preceding whitespaces.

Previously, we would just use `TextRange::empty(self.offset())` because the range for each token would be computed at the place where it was emitted. Now, the lexer handles this centrally in `next_token` which means we require special handling at certain places. This is one such case.

Computing the range centrally in `next_token` has a benefit that the return type of lexer methods is a simple `TokenKind` but it does add a small complexity to handle such cases. I think it's worth doing this but if anyone thinks otherwise, feel free to comment.

## Test Plan

Add a test case and update the snapshot.

### CODE REVIEW

File: crates/ruff_python_parser/src/lexer.rs [hunk 1]
```diff
@@ -994,6 +994,21 @@ impl<'src> Lexer<'src> {
...
+                // if first:
+                //     if second:
+                //         pass
+                //     foo
+                // #   ^
+                // ```
+                //
+                // Here, the cursor is at `^` and the `indentation` contains the whitespaces before
+                // the `pass` token.
+                self.cursor.start_token();
```
Comment (MichaReiser) [inline crates/ruff_python_parser/src/lexer.rs:1010]: I must admit that reading `self.cursor.start_token()` is hard to understand. Why are we starting a token when we're just about to return it. Your comment certainly helps. That makes me wonder if we can instead add a method on `Lexer` with a more meaningful name. 

```rust
self.update_token_start_to_current_offset() // A bi tlong
self.reset_token_start()
```

but I must admit, I don't find them much better than what we have. But maybe you have an idea for a better name?

**Review by MichaReiser — COMMENTED** [review]
There are in total 5 usages of `TextRange::empty(offset))` in the existing `Lexer` code. Would you mind reviewing if we need to make the same change in other places?

**Comment by dhruvmanila** [comment]
> There are in total 5 usages of `TextRange::empty(offset))` in the existing `Lexer` code. Would you mind reviewing if we need to make the same change in other places?

I did review them, sorry for not mentioning in the PR description. This isn't needed in all places, it's only require if the lexer skips whitespaces. For posterity, the usages are:

https://github.com/astral-sh/ruff/blob/685d11a909451731dcaf4e1284d0f9143efae6e8/crates/ruff_python_parser/src/lexer.rs#L853-L863

⬆️ This emits any pending dedent tokens. There are no whitespaces consumed by the lexer before this, so no need to reset the cursor.

https://github.com/astral-sh/ruff/blob/685d11a909451731dcaf4e1284d0f9143efae6e8/crates/ruff_python_parser/src/lexer.rs#L1031-L1041

⬆️ We already reset the cursor before this function is called: 

https://github.com/astral-sh/ruff/blob/b600b4287691ba0f95d63860957a895757e7053a/crates/ruff_python_parser/src/lexer.rs#L877-L898

File: crates/ruff_python_parser/src/lexer.rs (hunk 1 above)
Comment (dhruvmanila) [reply crates/ruff_python_parser/src/lexer.rs:1010]: Yeah, I agree. Let me think about this.

**Review by MichaReiser — APPROVED** [review]
Thanks for explaining!

**Pull request merged**